### PR TITLE
More efficient executable allocator

### PR DESCRIPTION
**Pull request by davidwrighton**
Improve the performance of the `ExecutableAllocator` via a series of changes which collectively reduce number of cache misses within the `ExecutableAllocator::MapRW` function and reduce the cost of such cache misses.

Adjustments to improve caching of RW pages:
- Add a concept of caching with multiple entries. This cache is fairly easy to configure by adjusting the size of the `ExecutableAllocator::m_cachedMapping[X]` array. In this PR based on experimentation with various different sizes of cache, I have chosen a cache size of 3 which shows substantial improvement from the previous cache size of 1. This cache is somewhat less effective in heavily multithreaded applications, but still works substantially better than a cache of 1.
- The various VSD stubs are move from a collection of `LoaderHeap` objects associated with the VirtualStubManager to using the `CodeFragmentHeap`. This has 2 benefits. The first of which is that stub type identification is now substantially faster for large applications, as it is able to use the `RangeSectionMap` to compute the stub kind instead of walking a singly linked list AKA the `RangeList`, and that for relatively small applications, the same page used to hold recently modified jitted code will also be used to hold recently allocated VSD stubs. This will reduce the number of different mappings that `MapRW` needs to process.
- The size of page size used for interleaved `LoaderHeap` structures was raised to a minimum of 16KB from the 4KB that was used. This reduces the count of mappings for this sort of stub by a factor of 4, and makes it reasonable to skip caching the RW mapping for the interleaved structures, which reduces the load on the RW mapping cache.

Cache miss rates...
| Application | Original cache miss count | New cache miss count |
| --- | --- | --- |
| Powershell | 1967 | 154 |
| Crossgen2 | 2172 | 428 |
| ASP.NET MVC template app | 2169 | 138 |

Improvement to the performance of cache misses
- The set of mappable regions in the `ExecutableAllocator` is organized as a singly linked list. In the change I made a very simple tweak to move the found region to the front of the list. This has resulted in substantial drop in the count of mappings walked to find the right one. Effectively, this treats the entire list of mappable regions as a MRU optimized list, which turns out to be a good assumption.

| Application | Original average walk length | New walk length |
| --- | --- | --- |
| Powershell | 103.3 | 3.3 |
| Crossgen2 | 74.5 | 2.4 |
| ASP.NET MVC template app | 101.3 | 2.4 |

### CODE REVIEW

File: src/coreclr/vm/virtualcallstub.cpp
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/virtualcallstub.cpp:568]: Some clarifying comments on the value `2` would be helpful. Is it possible we can compute this based on the `GetOsPageSize()` or is this simply a value we use for an optimization? We converted from 4k to 16k, factor of 4, but we changed this by a factor 3 so that is where I am coming from.

File: src/coreclr/inc/loaderheap.h
Comment (davidwrighton) [reply src/coreclr/inc/loaderheap.h:163]: ARM is special as the instruction set does not let you have a simple immediate offset that is 16KB in length. This makes it impractical to use a 16KB offset as it causes measurable perf problems to add the extra instructions necessary to handle a 16KB offset.

File: src/coreclr/utilcode/executableallocator.cpp
Comment (davidwrighton) [reply src/coreclr/utilcode/executableallocator.cpp:363]: No. You can make an assert with a Crst by checking OwnedByCurrentThread(), but CRITSEC_COOKIE doesn't have that nicety easily accessible.

File: src/coreclr/vm/loongarch64/thunktemplates.S
Comment (davidwrighton) [reply src/coreclr/vm/loongarch64/thunktemplates.S:27]: We don't build this code, and frankly its the next best thing to impossible to predict if any given assembler will accept input that makes sense. I think this is likely to work, but it is a best guess.

File: src/coreclr/vm/virtualcallstub.cpp
Comment (davidwrighton) [reply src/coreclr/vm/virtualcallstub.cpp:568]: This is about going from having 6 LoaderHeap structures as part of a VirtualCallStubManager to only having 2.

File: src/coreclr/utilcode/executableallocator.cpp
Comment (davidwrighton) [reply src/coreclr/utilcode/executableallocator.cpp:239]: Yeah. I was just lazy and didn't need those.. but it should probably be joined up with them... especially as the scheme I wrote doesn't really work in all environments.

File: src/coreclr/utilcode/executableallocator.cpp
Comment (janvorli) [inline src/coreclr/utilcode/executableallocator.cpp:363]: It is missing adding the refcount when you return here.

File: src/coreclr/vm/i386/thunktemplates.S
Comment (davidwrighton) [reply src/coreclr/vm/i386/thunktemplates.S:8]: My understanding was that x86 doesn't benefit from keeping the 4096 page size, as it only runs on Windows which has 64KB allocation granularity for memory reservation in the ExecutableAllocator in any case, thus resulting in no change to the available VM space.

File: src/coreclr/utilcode/loaderheap.cpp
Comment (davidwrighton) [reply src/coreclr/utilcode/loaderheap.cpp:1348]: Agreed.   I misunderstood the meaning of m_dwGranularity in one of my earlier changes and missed changing this one back.

File: src/coreclr/utilcode/loaderheap.cpp
Comment (davidwrighton) [reply src/coreclr/utilcode/loaderheap.cpp:1323]: Agreed.   I misunderstood the meaning of m_dwGranularity in one of my earlier changes and missed changing this one back.

File: src/coreclr/utilcode/executableallocator.cpp
Comment (davidwrighton) [reply src/coreclr/utilcode/executableallocator.cpp:363]: It doesn't need a ref count in this situation. In this situation it is finding the value in the cache, and like the old code doesn't need to adjust any ref counts.

File: src/coreclr/vm/i386/thunktemplates.S
Comment (janvorli) [reply src/coreclr/vm/i386/thunktemplates.S:8]: But we also have Linux x86 (not official, but used by the Samsung folks). As for windows x86, I was somehow under the impression that it was not using the 64kB granularity, but I can see I was wrong. However, larger page means potentially wasting more physical memory (if you allocate 16kB page and end up allocating just a fraction of it). Not sure if it is worth it, but I'll leave it up to you.

File: src/coreclr/utilcode/executableallocator.cpp
Comment (janvorli) [reply src/coreclr/utilcode/executableallocator.cpp:363]: I am sorry, you are right, I've misread my original code.

File: src/coreclr/inc/executableallocator.h
Comment (janvorli) [inline src/coreclr/inc/executableallocator.h:132]: David, I can see some coreclr tests in the CI failing on x64 macOS with illegal instruction, so maybe the failure you are seeing is unrelated to your changes. This is from my recent PR:
```
/private/tmp/helix/working/A18108BE/w/B44209DD/e /private/tmp/helix/working/A18108BE/w/B44209DD/e
  Discovering: System.Runtime.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.Runtime.Tests (found 9140 of 9186 test cases)
  Starting:    System.Runtime.Tests (parallel test collections = on, max threads = 4)
./RunTests.sh: line 168: 76212 Illegal instruction: 4  "$RUNTIME_PATH/dotnet" exec --runtimeconfig System.Runtime.Tests.runtimeconfig.json --depsfile System.Runtime.Tests.deps.json xunit.console.dll System.Runtime.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=AdditionalTimezoneChecks -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing $RSP_FILE
/private/tmp/helix/working/A18108BE/w/B44209DD/e
----- end Wed Mar 29 16:33:56 EDT 2023 ----- exit code 132 ----------------------------------------------------------
exit code 132 means SIGILL Illegal Instruction. Core dumped. Likely codegen issue.
```

Comment (davidwrighton) [reply src/coreclr/inc/executableallocator.h]: Seems likely. I've reverted my cache size tweak. @janvorli could you review this change and sign off? I'm looking to check this in today/Monday. If you think I need someone else to signoff on the type system stuff, please let me know.

**Comment by janvorli** [comment]
@davidwrighton it looks good to me except for the https://github.com/dotnet/runtime/pull/83632#discussion_r1154889452

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

File: src/coreclr/vm/loongarch64/thunktemplates.S
Comment (shushanhf) [reply src/coreclr/vm/loongarch64/thunktemplates.S:29]: @AaronRobinsonMSFT   @davidwrighton 
This is not right for LA64.
I will revert the modification for LA64.

Comment (shushanhf) [reply src/coreclr/vm/loongarch64/thunktemplates.S:29]: For LA64, the OS-PageSize is 16k, the  `pcaddi  $t2, 0x1000`  is right.

**Pull request merged**
